Training/python/2018v1/Training_v0p1.py

```diff
@@ -267,16 +295,42 @@ def compile_model(model, opt_name, learning_rate):
 
 def run_training(model, data_loader, to_profile, log_suffix):
 
-    gen_train = data_loader.get_generator(primary_set = True, return_weights = data_loader.use_weights)
-    gen_val = data_loader.get_generator(primary_set = False, return_weights = data_loader.use_weights)
-    input_shape, input_types = data_loader.get_input_config()
-
-    data_train = tf.data.Dataset.from_generator(
-        gen_train, output_types = input_types, output_shapes = input_shape
-        ).prefetch(tf.data.AUTOTUNE)
-    data_val = tf.data.Dataset.from_generator(
-        gen_val, output_types = input_types, output_shapes = input_shape
-        ).prefetch(tf.data.AUTOTUNE)
+    if data_loader.input_type == "tf":
+        total_batches = data_loader.n_batches + data_loader.n_batches_val
+        tf_dataset_x_order = data_loader.tf_dataset_x_order
+        tauflat_index = tf_dataset_x_order.index("TauFlat")
+        inner_indices = [i for i, elem in enumerate(tf_dataset_x_order) if 'inner' in elem]
+        outer_indices = [i for i, elem in enumerate(tf_dataset_x_order) if 'outer' in elem]
+        ds = tf.data.experimental.load(data_loader.tf_input_dir, compression="GZIP") # import dataset
+        if data_loader.rm_inner_from_outer:
+            my_ds = ds.map(lambda x, y, weights: rm_inner(x, y, weights, outer_indices))
+        else: 
+            my_ds = ds
+        cell_locations = data_loader.cell_locations
+        active_features = data_loader.active_features
+        active = [] #list of elements to be kept
+        if "TauFlat" in active_features:
+            active.append(tauflat_index)
+        if "inner" in cell_locations:
+            active.extend(inner_indices)
+        if "outer" in cell_locations:
+            active.extend(outer_indices)
```

Comment:
To my understanding the model input shape is pre-defined in the `netConf_full` variable, you can not just change the number of tensors in the input (to NN) tuple because it already considers fixed number and defined here: https://github.com/cms-tau-pog/TauMLTools/blob/ed7b7b385219fa6a239bf972502cb8edf92754d1/Training/python/DataLoader.py#L185-L219

Also my question related to the indexes of the tensor (em, muon, hadrons,...) was referring that the sequential number of this index is already defined in meta enume class (see attachments) [SetupMeta.txt](https://github.com/cms-tau-pog/TauMLTools/files/8237831/SetupMeta.txt)
if you at some point need this index, maybe it would be better by designed not to define `tf_dataset_x_order` but take it from `enum class CellObjectType {` for instance `getattr(R.CellObjectType,"PfCand_electron")` would give you 1 or if you need the order of the grouped tensors after DataLoading it is following this order: https://github.com/cms-tau-pog/TauMLTools/blob/ed7b7b385219fa6a239bf972502cb8edf92754d1/Training/configs/training_v1.yaml#L72-L76 if you remove from there one of the list it should not return this tensor. But in principle this part is not interfering with RUN3 training, so it is up to you.

Reply:
From the testing I've done, changing the number of inputs in the tensors works if you change the parameters that I introduced to switch off either inner or outer cones (cell_locations) or the tau dense layer (removing it from active_features), this information is passed to net config. We then need to remove the concerned layers from the saved tensor if using tf input format as they are usually created with all layers active (eg generated with all 7 tensors 3xinner, 3xouter and tau dense, so if we want to then disable for example inner for a training from this tf dataset we need to remove the 3 inner tensors from the dataset as the model will only expect 4 tensors). Let me know if this doesn't make sense.

Reply:
Will also have a look at the indexes you're referring to in SetupMeta, thanks!